src/runner_metrics.py

```diff
@@ -0,0 +1,198 @@
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s.", PRE_JOB_METRICS_FILE_NAME, fs.runner_name)
+        return None
+
+    try:
+        post_job_metrics = json.loads(fs.path.joinpath(POST_JOB_METRICS_FILE_NAME).read_text())
+        logger.debug("Post-job metrics for runner %s: %s", fs.runner_name, post_job_metrics)
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s", POST_JOB_METRICS_FILE_NAME, fs.runner_name)
+        post_job_metrics = None
+
+    return RunnerMetrics(
+        installed_timestamp=installed_timestamp,
+        pre_job=PreJobMetrics(**pre_job_metrics),
+        post_job=PostJobMetrics(**post_job_metrics) if post_job_metrics else None,
+    )
+
+
+def _issue_runner_metrics(runner_metrics: RunnerMetrics, flavor: str) -> None:
+    """Issue metrics.
+
+    Converts the metrics into respective metric events and issues them.
+
+    Args:
+        runner_metrics: The metrics to be issued.
+        flavor: The flavor of the runners.
+    """
+    event = metrics.RunnerStart(
+        timestamp=runner_metrics.pre_job.timestamp,
+        flavor=flavor,
+        workflow=runner_metrics.pre_job.workflow,
+        repo=runner_metrics.pre_job.repository,
+        github_event=runner_metrics.pre_job.event,
+        idle=runner_metrics.pre_job.timestamp - runner_metrics.installed_timestamp,
+    )
+    metrics.issue_event(event)
+
+
+def _clean_up_shared_fs(fs: shared_fs.SharedFilesystem) -> None:
+    """Clean up the shared filesystem.
+
+    Remove all metric files and afterwards the shared filesystem.
+    Args:
+        fs: The shared filesystem for a specific runner.
+    """
+    fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME).unlink(missing_ok=True)
+    fs.path.joinpath(POST_JOB_METRICS_FILE_NAME).unlink(missing_ok=True)
+    fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME).unlink(missing_ok=True)
+
+    shared_fs.delete(fs.runner_name)
+
+
+def extract(flavor: str, ignore_runners: set[str]) -> None:
+    """Extract and issue metrics from runners.
+
+    The metrics are extracted from the shared filesystem of given runners
+    and respective metric events are issued.
+    Orphan shared filesystems are cleaned up.
+
+    If corrupt data is found, an error is raised immediately, as this may indicate that a malicious
+    runner is trying to manipulate the shared file system.
+    In order to avoid DoS attacks, the file size is also checked.
+
+    Args:
+        flavor: The flavor of the runners to extract metrics from.
+        ignore_runners: The set of runners to ignore.
+
+    Raises:
+        CorruptDataError: If one of the files inside the shared filesystem is not valid.
+    """
+    for fs in shared_fs.list_all():
+        if fs.runner_name not in ignore_runners:
+            try:
+                metrics_from_fs = _extract_metrics_from_fs(fs)
+            except (JSONDecodeError, ValidationError, MetricFileSizeTooLargeError) as exc:
```

Comment:
It would be good for `_extract_metrics_from_fs` not to bubble up errors like `JSONDecodeError` or any others not explicitly raised in `_extract_metrics_from_fs` just so that it is clear what exceptions can be raised in each function and to avoid having implementation details leak to the caller